### ISSUE TITLE
Potential regression for contrib (backup_migrate) caused by initial php 8 support

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#5094]
**Description of the bug**

Several users reported a problem with backup_migrate after updating to Backdrop 1.19.

https://github.com/backdrop-contrib/backup_migrate/issues/63

`Error : Cannot access protected property DatabaseConnection_mysql::$pdo`

`Error : Call to undefined method DatabaseConnection_mysql::exec()`

**Steps to reproduce:**
* Install backdrop 1.19.0
* Install and enable backup and migrate 1.x-1.0.16
* Attempt to create a backup
* Attempt to restore from this backup

---

I think, the first thing people need is instructions on how to fix this on their sites:

To avoid the problem, update Backup Migrate to 1.x-1.0.17 *before* you update core to 1.19.0

OR

If you're already in trouble, comment out one line in Backup Migrate as described [in this issue comment](https://github.com/backdrop-contrib/backup_migrate/issues/63#issuecomment-842397222) (then update the module)

**Additional information**

- Backdrop CMS version: 1.19.0
- Backup Migrate versions before 1.x-1.0.17

Probably caused by https://github.com/backdrop/backdrop/commit/4102cd21190e75e35e4274472565691f7e4fb6a9#diff-a30038705d970e43b622562028439144c59afc2abc9fb53048aff64286fbcfc3

**Comment by olafgrabienski** [comment]
In the meantime, more reports came up, amongst others https://github.com/backdrop/backdrop-issues/issues/5096 with three suggestions how to handle the situation:

- Issue some kind of notification warning users that they need to upgrade Backup and Migrate module if they use 1.19.
- Release 1.19.1 with some fix that remediates the problems with Backup and Migrate.
- Assume that users that update Backdrop Core after today will also be updating Backup and Migrate, just because they see it's available, and that this won't be an issue moving forward?

**Comment by indigoxela** [comment]
> Release 1.19.1 with some fix that remediates the problems with Backup and Migrate

I don't think that there's an easy fix, I might be wrong, though.

> Assume that users that update Backdrop Core after today will also be updating Backup and Migrate

What worries me about that is that users, who aren't on GitHub (or follow on Twitter) will not be aware of the problem - and it's a fatal one.

**Comment by klonos** [comment]
Just chiming in to say that https://backdropcms.org/project/usage/backup_migrate reports 650+ installation of the module. So it seems like ~40% of Backdrop sites have it installed, and will potentially be effected by this.

PS: if we check again in a week or so, and the reported usage shows most sites already having switched to v1.x-1.0.16 of the module, then perhaps we don't need to do anything 🤷🏼 ...unless this turns out to be an easy fix of course.

**Comment by indigoxela** [comment]
> it seems like ~40% of Backdrop sites have it installed

Yes, that's potentially a lot of people running into problems ... or maybe not, if they update the module in time.

@klonos regarding the usage pages: unless you have access to uncached / more recent data, these pages are worthless. The usage stats there usually lag behind two weeks, so I wouldn't recommend to rely on them.

**Comment by stpaultim** [comment]
My issue was a duplicate. As I think about it, I think I opened "Problem with Backup and Migrate Module - conflict with 1.19.0" - because I wanted an issue with a very obvious name that might catch the attention of anyone in trouble with this problem. 

I think that the title of this issue might be overlooked by someone grappling with this issue. But, this is not important. 

I've been posting notices about the importance of updating Backup and Migrate anywhere I can think of. 

- I opened an issue in the forum with a very clear title. 
https://forum.backdropcms.org/forum/problem-backup-and-migrate-and-backdrop-1190
- I posted to twitter - https://twitter.com/backdropcms/status/1394506087081070594
- I added a message to the dashboard notifications
![dashboard](https://user-images.githubusercontent.com/3144571/118751299-b16e5c80-b826-11eb-86ed-213d2d72f5c7.png)
- We could mention this issue in our blog post for 1.19.0, but we don't have one yet. Maybe, I'll start one and mention this problem there.

**Comment by stpaultim** [comment]
@jenlampton - asked in Zulip about possible language for the release notes:

How about:

> The Backdrop community is proud to release version 1.19.0 of Backdrop CMS, following our 4-month release cycle.
> 
> This release contains several new features, user experience improvements, and a few bug fixes for these recent improvements.
> 
> Some users have reported a conflict with earlier versions of the Backup and Migrate module, please be sure to upgrade to the latest version of Backup and Migrate ASAP.

**Comment by indigoxela** [comment]
> Some users have reported a conflict with earlier versions of the Backup and Migrate module, please be sure to upgrade to the latest version of Backup and Migrate ASAP.

One minor suggestion: also mention the exact version 1.x-1.0.17. Because of update data caching, it might take a little time, until the most recent version shows up in the list - depending on the update check frequency setting.

And possibly a little shorter? But that hint is for sure helpful. :+1:

**Comment by indigoxela** [comment]
FTR: I did some testing yesterday - it seems to be OK to update both, core and module in the _same_ update run via UI.

@stpaultim which changes to this issue's title would you recommend to make it easier to discover?

**Comment by jenlampton** [comment]
I have added the following to the release notes:

`Please update Backup and Migrate to 1.x-1.0.17 along with this core update. Some people have reported a conflict with earlier versions of the module.`

**Comment by quackers8me** [comment]
A related issue is that due to the PHP 8 changes, you cannot run restores from backup_migrate backdrop-contrib/backup_migrate#64 
Restoring from the command line from a backup made with backup_migrate still works.

The issue may not be able to be solved without a change to B 1.19 core database.inc because it no longer allows to call prepare(), only prepareQuery() which strips out `{}` from a backup file.

The D7 dev branch of database.inc which has changes related to PHP 8 has the following code that may work:

`  public function __call($name, $arguments) {
    return call_user_func_array(array($this->connection, $name), $arguments);
  }`

**Comment by stpaultim** [comment]
Given that the Backup and Migrate module is the most popular contrib module for Backdrop CMS and that without "restore" functionality, it's value is greatly reduced to users - I'd like to put this issue on the agenda for our dev meeting tomorrow.

**Comment by stpaultim** [comment]
This was discussed during the dev meeting today, see recording of meeting: https://youtu.be/_W_s1yt8fFs. The discussion of this issue takes place at the 16:50 minute mark. 

It was my understanding that @quicksketch has ideas on how to proceed and that he plans to update this issue.

**Comment by quackers8me** [comment]
I watched the dev meeting and just to add colour to the issue here are the
details of events:
When B 1.19 was released, the removal of “extends PDO” from database.inc
caused a prepared statement to fail. This caused a complete site crash and
an error message related to backup_migrate to display.
In speaking with the D7 module maintainer, this code in question was
removed recently due to a MySQL 8 issue (not PHP 8 because D7 hasn’t
proceeded there yet).

The backup_migrate module was updated and 1.17 was released.

Later user reports found that the restore function failed and the only way
to restore was from the command line.

I have reached out a second time to the D7 module maintainer and have
not heard back regarding the restore issue.

**Comment by jenlampton** [comment]
@indigoxela  I added a guess at steps to reproduce this problem in the Original Post, but do you remember at which step the errors would appear? Has anyone else

**Comment by quicksketch** [comment]
The D7 dev branch of database.inc which has changes related to PHP 8 has the following code that may work:

```
public function __call($name, $arguments) { return call_user_func_array(array($this->connection, $name), $arguments); }
```

This approach would probably work, though I strongly dislike magic methods like this because IDEs cannot follow the call stack through them. It also makes documentation nearly impossible because there is no place to put docblocks. This results in neither machines nor humans being able to figure out what methods are available in an object unless you know it's there.

Although it's not as comprehensive a fix as adding a magic method, I filed a PR at https://github.com/backdrop/backdrop/pull/3646 that solves the problem specifically for Backup and Migrate, by simply wrapping the `PDO::prepare()` method manually in a new `DatabaseConnection::prepare()` method.

In testing, this seems to solve the problem with Backup and Migrate and I can do a full site restoration.

**Comment by quicksketch** [comment]
If we wanted to comprehensively use this approach, we'd also need to wrap the following methods (from [the PDO documentation](https://www.php.net/PDO)):

- PDO::beginTransaction — Initiates a transaction
- PDO::commit — Commits a transaction
- PDO::errorCode — Fetch the SQLSTATE associated with the last operation on the database handle
- PDO::errorInfo — Fetch extended error information associated with the last operation on the database handle
- PDO::exec — Execute an SQL statement and return the number of affected rows
- PDO::getAttribute — Retrieve a database connection attribute
- PDO::getAvailableDrivers — Return an array of available PDO drivers
- PDO::inTransaction — Checks if inside a transaction
- PDO::lastInsertId — Returns the ID of the last inserted row or sequence value
- PDO::prepare — Prepares a statement for execution and returns a statement object
- PDO::query — Executes an SQL statement, returning a result set as a PDOStatement object
- PDO::quote — Quotes a string for use in a query
- PDO::rollBack — Rolls back a transaction
- PDO::setAttribute — Set an attribute

But the PR seems to fix Backup and Migrate specifically.

**Comment by indigoxela** [comment]
@quicksketch many thanks for the PR, I've left a comment regarding wording.

How should this get tested? Is verifying, that the BaM restore works correctly sufficient?

**Comment by quicksketch** [comment]
> How should this get tested? Is verifying, that the BaM restore works correctly sufficient?

Correct. To test:

* Install Backup and Migrate module
* Create a backup at `/admin/config/system/backup_migrate`, Download the file
* Restore the backup by uploading the file at `/admin/config/system/backup_migrate/restore`
* Confirm that no errors occurred.

**Comment by quackers8me** [comment]
I can confirm that the changes in the PR works correctly with the Backup
and Migrate module.

**Comment by indigoxela** [comment]
@quackers8me many thanks for confirming.

I also did a quick check in the sandbox, these were my steps:

- Install Backup and Migrate
- Create a private file system
- Took a backup
- Edited the first post (added some text)
- Restored from the backup
- Verified that there were no errors and the first post has the initial text again

Also works for me. :+1:

**Comment by indigoxela** [comment]
Regarding the approach - pass-through vs. magic method in D7 - I agree, that this is more comprehensive and that a proper doc block has high value.
I'm no expert regarding PDO or the Backdrop database abstraction layer in general, so maybe a second pair of eyes to review the code would be helpful to identify potential side effects (if any).

But people are waiting for a fix, so they can safely use BaM again - a backup is only useful, it you can actually restore. Tentatively marking as RTBC.

Regarding all the other methods and if the abstraction layer should provide pass-through for all of those ... maybe a new issue makes sense? This might be a high-level discussion.

**Comment by stpaultim** [comment]
Just to confirm that people are waiting on a solution. Here is a link to a comment posted today in the forum: https://forum.backdropcms.org/comment/2907#comment-2907

**Comment by klonos** [comment]
> Regarding all the other methods and if the abstraction layer should provide pass-through for all of those ... maybe a new issue makes sense?

Follow-up issue created here: #5120 😉

**Comment by indigoxela** [comment]
People are really waiting for this, so let's try to get some progress. :wink:

**Comment by quicksketch** [comment]
Thanks folks! I have merged this into 1.x and 1.19.x. I'll shoot for getting 1.19.2 out this week.
